measured.go

```diff
@@ -380,74 +263,35 @@ func (m *Measured) newConn(c net.Conn, interval time.Duration) net.Conn {
 // Read() implements the function from net.Conn
 func (mc *Conn) Read(b []byte) (n int, err error) {
 	n, err = mc.Conn.Read(b)
-	if err != nil {
-		mc.submitError(err, "read")
-	}
 	atomic.AddUint64(&mc.BytesIn, uint64(n))
 	return
 }
 
 // Write() implements the function from net.Conn
 func (mc *Conn) Write(b []byte) (n int, err error) {
 	n, err = mc.Conn.Write(b)
-	if err != nil {
-		mc.submitError(err, "write")
-	}
 	atomic.AddUint64(&mc.BytesOut, uint64(n))
 	return
 }
 
 // Close implements the function from net.Conn
 func (mc *Conn) Close() (err error) {
 	err = mc.Conn.Close()
-	if err != nil {
-		mc.submitError(err, "close")
-	}
 	mc.submitTraffic()
 	mc.chStop <- struct{}{}
 	return
 }
 
-func (mc *Conn) submitError(err error, phase string) {
-	mc.m.submitError(mc.ID, err, phase)
-}
-
 func (mc *Conn) submitTraffic() {
 	mc.m.submitTraffic(mc.ID,
 		atomic.SwapUint64(&mc.BytesIn, 0),
 		atomic.SwapUint64(&mc.BytesOut, 0))
 }
 
-func (m *Measured) submitError(connID string, err error, phase string) {
-	if atomic.LoadInt32(&m.stopped) == 1 {
-		log.Trace("Measured stopped, not submitting error")
-		return
-	}
-	splitted := strings.Split(err.Error(), ":")
-	lastIndex := len(splitted) - 1
-	if lastIndex < 0 {
-		lastIndex = 0
-	}
-	e := strings.Trim(splitted[lastIndex], " ")
-	er := &Error{
-		ID:    connID,
-		Error: e,
-		Phase: phase,
-	}
-	log.Tracef("Submitting error %+v", er)
-	m.chStat <- er
-}
-
-func (m *Measured) submitTraffic(connID string, BytesIn uint64, BytesOut uint64) {
+func (m *Measured) submitTraffic(connID string, in uint64, out uint64) {
 	if atomic.LoadInt32(&m.stopped) == 1 {
 		log.Trace("Measured stopped, not submitting traffic")
 		return
 	}
-	t := &Traffic{
-		ID:       connID,
-		BytesIn:  BytesIn,
-		BytesOut: BytesOut,
-	}
-	log.Tracef("Submitting traffic %+v", t)
-	m.chStat <- t
+	m.trackTraffic(connID, in, out)
```

Comment:
Rather than submitting traffic objects to a channel, I just call trackTraffic directly, which does the summarizing calculation immediately.  This does involve locking on a mutex, but the calculation should be really fast so I'm not concerned about this.